### PR TITLE
Re-implement the prefetcher using backpressure mechanism

### PR DESCRIPTION
**Pull request by monthonk**
## Description of change

The prefetcher now uses only one GetObject request to fetch data in advance. This request has a range of entire object but use backpressure mechanism to control how much data it wants to fetch into the part queue instead of spawning up to two requests in parallel.

This should make the throughput more stable because previously the two request tasks could compete with each other when fetching data from S3. Also, it will be easier to control how much data we want to store in the part queue.

## Does this change impact existing behavior?

No, the change only for internal interface.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/prefetch/part_stream.rs [hunk 1]
```diff
@@ -223,12 +227,35 @@ where
...
                                 curr_offset += part.len() as u64;
-                                part_queue_producer.push(Ok(part));
+                                part_queue_producer.push(Ok(part), Some(next_read_window_offset));
+                            }
+                            debug_assert!(
+                                curr_offset <= last_offset,
+                                "current offset must not be larger than last offset"
+                            );
+
+                            if remaining_window_size > 0 {
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/part_stream.rs:237]: It's not clear to me why we are calling `increment_read_window` here or why we need the sender/receiver pair. I'd argue this logic requires some comment or maybe a rethink.

File: mountpoint-s3/src/prefetch/task.rs [hunk 2]
```diff
@@ -14,36 +15,62 @@ pub struct RequestTask<E: std::error::Error> {
...
             start_offset: offset,
             total_size: size,
             part_queue,
+            read_window_updater,
         }
     }
 
+    // Push a given list of parts in front of the part queue
+    pub async fn push_front(&mut self, mut parts: Vec<Part>) -> Result<(), PrefetchReadError<E>> {
+        // Merge all parts into one single part.
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/task.rs:58]: This could be expensive (especially because of checksumming). Couldn't we just keep a vector of parts?

File: mountpoint-s3/src/prefetch/part_stream.rs [hunk 3]
```diff
@@ -168,49 +169,52 @@ where
...
 
-                pin_mut!(get_object_result);
+                pin_mut!(request);
                 loop {
-                    match get_object_result.next().await {
+                    match request.next().await {
                         Some(Ok((offset, body))) => {
                             trace!(offset, length = body.len(), "received GetObject part");
-                            metrics::counter!("s3.client.total_bytes", "type" => "read").increment(body.len() as u64);
+
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/part_stream.rs:211]: Are we not reporting `total_bytes` anymore?

**Review by passaro — COMMENTED** [review]
Not a full review, but I do have a few questions before continuing.

File: mountpoint-s3/src/prefetch.rs [hunk 4]
```diff
@@ -110,35 +112,50 @@ where
...
-            // This is a weird looking number! We really want our first request size to be 1MiB,
-            // which is a common IO size. But Linux's readahead will try to read an extra 128k on on
-            // top of a 1MiB read, which we'd have to wait for a second request to service. Because
-            // FUSE doesn't know the difference between regular reads and readahead reads, it will
-            // send us a READ request for that 128k, so we'll have to block waiting for it even if
-            // the application doesn't want it. This is all in the noise for sequential IO, but
-            // waiting for the readahead hurts random IO. So we add 128k to the first request size
-            // to avoid the latency hit of the second request.
-            first_request_size: 1 * 1024 * 1024 + 128 * 1024,
+            initial_read_window_size,
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch.rs:158]: Yes, that's right. The CRT will return a part size anyway even if the read window is smaller, so the change here should reflect that. The latency could also be impacted as the first request size is larger.

File: mountpoint-s3/src/prefetch.rs [hunk 5]
```diff
@@ -284,31 +323,47 @@ where
...
         while to_read > 0 {
-            let Some(current_task) = self.current_task.as_mut() else {
-                // If [prepare_requests] didn't spawn a request, we've reached the end of the object.
+            let Some(current_task) = self.backpressure_task.as_mut() else {
                 trace!(offset, length, "read beyond object size");
                 break;
             };
             debug_assert!(current_task.remaining() > 0);
 
+            if let Some(len) = self.next_read_window_size {
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch.rs:338]: I will try to consolidate it, probably easier if this logic is in `RequestTask`.

File: mountpoint-s3/src/prefetch.rs [hunk 6]
```diff
@@ -284,31 +323,47 @@ where
...
+                trace!(
+                    current_window = self.current_read_window_size,
+                    next_window = len,
+                    "incrementing read window"
+                );
+                current_task.increment_read_window(len);
+                self.current_read_window_size = len;
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch.rs:345]: It actually remembers the last window size we have increased. Probably bad choice of naming here.

File: mountpoint-s3/src/prefetch.rs [hunk 7]
```diff
@@ -458,61 +469,45 @@ where
...
+        if offset >= available_soon_offset {
             trace!(
                 requested_offset = offset,
                 available_offset = available_offset,
                 "seek failed: not enough data available"
             );
             return Ok(false);
         }
+
+        // Make sure there is enough data in the read window
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch.rs:491]: I think you're right, will update this.

File: mountpoint-s3/src/prefetch/task.rs (hunk 2 above)
Comment (monthonk) [reply mountpoint-s3/src/prefetch/task.rs:58]: My thinking is that it's fairly cheap compare to creating a new request and this way we don't have to change the part queue structure. I will see if we can improve this.

File: mountpoint-s3/src/prefetch.rs [hunk 8]
```diff
@@ -206,22 +248,19 @@ pub struct PrefetchGetObject<Stream: ObjectPartStream, Client: ObjectClient> {
...
+    backpressure_task: Option<RequestTask<Client::ClientError>>,
     // Invariant: the offset of the last byte in this window is always
     // self.next_sequential_read_offset - 1.
     backward_seek_window: SeekWindow,
     bucket: String,
     object_id: ObjectId,
     // preferred part size in the prefetcher's part queue, not the object part
     preferred_part_size: usize,
+    current_read_window_size: usize,
+    next_read_window_size: Option<usize>,
```
Comment (vladem) [inline mountpoint-s3/src/prefetch.rs:260]: May it make things simpler if we'd always reason in terms of absolute offsets, including this `next_read_window_size` variable?

I understand that now you use `next_read_window_size` as a relative value, which will be added to the current allowed offset. The problem is that since we mix absolute offsets and relative values, the code is hard to follow (e.g. you need to look around to understand what's `fn prepare_next_read_window` is doing): you always need to keep in mind what is relative and what is absolute. 

May be if we'd just record the last allowed offset that we've set for the request and the previous one, we'd still be able to calculate the value to use in `current_task.increment_read_window(..)`?

As aside -- didn't get why the calculation of `next_read_window_size` is separated from the call to `increment_read_window`. This creates the need for this `PrefetchGetObject::next_read_window_size` field, which could've been just a local variable.

File: mountpoint-s3/src/prefetch.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (vladem) [reply mountpoint-s3/src/prefetch.rs:491]: +1 for resetting if `next_read_window_offset < available_soon_offset`.

So `max_forward_seek_wait_distance` [today](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3/src/prefetch.rs#L149) is `16MB` and we assume that this is more efficient for the existing request to download `16MB` instead of creating a new one. Monthon's code looks aligned with this assumption (after line [482](https://github.com/awslabs/mountpoint-s3/pull/926/files#diff-ff2b2b24fbf3d7ad046bb857a44d84da70bbdd714b0ea9460f5d1b8a08408effR482) we know that it's less than `16MB` to be downloaded left to serve the read).

I'd say the problem here is that we may increment the window exactly to the requested byte and no further. Any window incrementing rule which prevents this looks arbitrary or complicated, so I'd just reset too.

File: mountpoint-s3/src/prefetch.rs [hunk 9]
```diff
@@ -284,31 +323,47 @@ where
...
+                trace!(
+                    current_window = self.current_read_window_size,
+                    next_window = len,
+                    "incrementing read window"
+                );
+                current_task.increment_read_window(len);
+                self.current_read_window_size = len;
+                self.next_read_window_size = None;
+            }
             let part = match current_task.read(to_read as usize).await {
```
Comment (vladem) [inline mountpoint-s3/src/prefetch.rs:348]: not clear how do we make sure that this will never block (which would mean a deadlock as this function is the only one increasing the window), I believe you guarantee that `to_read` is always in the window, a comment may help

File: mountpoint-s3/src/prefetch.rs [hunk 10]
```diff
@@ -354,54 +408,27 @@ where
...
-        let start = self.next_request_offset;
-        if start >= self.size {
-            return None;
-        }
+    /// Spawn a backpressure GetObject request which has a range from current offset to the end of the file.
+    /// We will be using flow-control window to control how much data we want to download into the prefetcher.
+    fn spawn_read_backpressure_request(&mut self) -> RequestTask<Client::ClientError> {
+        let start = self.next_sequential_read_offset;
+        let object_size = self.size as usize;
+        let range = RequestRange::new(object_size, start, object_size);
```
Comment (vladem) [inline mountpoint-s3/src/prefetch.rs:430]: dumb question, but we do have a test for backpressure-enabled requests, that they handle ranges well, i.e when request reads from the middle of an object?

File: mountpoint-s3/src/prefetch/part_queue.rs [hunk 11]
```diff
@@ -109,6 +137,9 @@ impl<E: std::error::Error + Send + Sync> PartQueueProducer<E> {
             trace!("closed channel");
         } else {
             self.bytes_sent.fetch_add(part_len, Ordering::SeqCst);
+            let _ = self
+                .next_read_window_offset
```
Comment (vladem) [inline mountpoint-s3/src/prefetch/part_queue.rs:141]: do we actually need to receive `next_read_window_offset` from the client? don't we have enough information in `Prefetcher` / `RequestTask` to compute that value? 

having to pass `next_read_window_offset` from `ObjectPartStream` to `Prefetcher` through multiple layers so it can just do:
```rust
let next_read_window_offset = current_task.next_read_window_offset();
self.prepare_next_read_window(remaining_window, next_read_window_offset);
```
looks overcomplicated, or am I missing something?

File: mountpoint-s3/src/prefetch.rs (hunk 10 above)
Comment (monthonk) [reply mountpoint-s3/src/prefetch.rs:421]: I think the tests that reset the prefetcher should cover this, like random read or forward seek, but I didn't add new tests specifically for it.

File: mountpoint-s3/src/prefetch/part_queue.rs (hunk 11 above)
Comment (monthonk) [reply mountpoint-s3/src/prefetch/part_queue.rs]: I just updated the code to simplify the flow. It should not be too complicated now.

File: mountpoint-s3-client/src/s3_crt_client/get_object.rs [hunk 12]
```diff
@@ -99,14 +107,23 @@ pub struct S3GetObjectRequest {
...
 
 impl GetObjectRequest for S3GetObjectRequest {
     type ClientError = S3RequestError;
 
     fn increment_read_window(mut self: Pin<&mut Self>, len: usize) {
+        self.next_read_window_offset += len as u64;
         self.request.meta_request.increment_read_window(len as u64);
     }
+
+    fn next_read_window_offset(self: Pin<&Self>) -> u64 {
```
Comment (passaro) [inline mountpoint-s3-client/src/s3_crt_client/get_object.rs:124]: This may require some comment or a better name. Why `next`? Isn't this the upper bound of the current read window?

File: mountpoint-s3/src/prefetch/backpressure_controller.rs [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    next_read_window_offset: u64,
+    last_request_offset: u64,
+}
+
+#[derive(Debug)]
+pub struct BackpressureLimiter {
+    read_window_incrementing_queue: Receiver<usize>,
+}
+
+/// Creates a [BackpressureController] and its related [BackpressureLimiter].
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/backpressure_controller.rs:19]: Could we explain what's the function of these 2 types? Their methods and fields could also do with some comments (and/or clearer names).

File: mountpoint-s3/src/prefetch/caching_stream.rs [hunk 14]
```diff
@@ -207,29 +262,65 @@ where
...
+                    let next_offset = offset + body.len() as u64;
+                    debug_assert!(
+                        next_offset <= block_aligned_byte_range.end,
+                        "current offset must be withing the block range"
+                    );
 
-                    // Split the body into blocks.
                     let mut body: Bytes = body.into();
+
+                    // S3 doesn't provide checksum for us if the request range is not aligned to
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/caching_stream.rs:273]: It is not clear to me what this whole block is doing.

File: mountpoint-s3/src/prefetch/caching_stream.rs [hunk 15]
```diff
@@ -183,7 +218,27 @@ where
             }
         };
 
-        pin_mut!(get_object_result);
+        pin_mut!(get_object_request);
+
+        // Current read window might be larger or smaller by client's read window at this point so we need to match them
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/caching_stream.rs:223]: This block needs some clarification. Why do we need this check? How do we match them?

File: mountpoint-s3/src/prefetch.rs [hunk 16]
```diff
@@ -347,7 +386,17 @@ where
...
 
             self.next_sequential_read_offset += part_bytes.len() as u64;
-            self.prepare_requests();
+
+            // Increase the read window size if available bytes in the queue are less than half of previous read window size
+            let remaining_in_queue = current_task
+                .available_offset()
+                .saturating_sub(self.next_sequential_read_offset) as usize;
+            if remaining_in_queue < self.preferred_read_window_size / 2 {
+                self.preferred_read_window_size = (self.preferred_read_window_size
```
Comment (passaro) [inline mountpoint-s3/src/prefetch.rs:395]: Do we need `self.preferred_read_window_size` now? Maybe it should belong to `RequestTask`, which could handle the read window increase when triggered here.

File: mountpoint-s3/src/prefetch/part_queue.rs [hunk 17]
```diff
@@ -93,15 +94,33 @@ impl<E: std::error::Error + Send + Sync> PartQueue<E> {
...
+
+        if let Some(current_part) = current_part.as_mut() {
+            part.extend(current_part).unwrap();
+            *current_part = part;
+        } else {
+            *current_part = Some(part);
+        }
+    }
+
+    pub fn bytes_received(&self) -> u64 {
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch/part_queue.rs:115]: I think `bytes_received` could be the size of the object now where previously it should not be larger than 2GB per task.

File: mountpoint-s3-client/src/s3_crt_client/get_object.rs (hunk 12 above)
Comment (monthonk) [reply mountpoint-s3-client/src/s3_crt_client/get_object.rs:124]: In my mind current offset is at the start of the read window (the lower bound), so the upper bound would be `next`. I will add the comment and also happy to change if it doesn't make sense.

File: mountpoint-s3/src/prefetch.rs (hunk 16 above)
Comment (monthonk) [reply mountpoint-s3/src/prefetch.rs:395]: I'm not sure where it should be either, but I believe the prefetcher should be in control of scaling up or down of the read window size. That's why I keep it here.

Comment (vladem) [reply mountpoint-s3/src/prefetch.rs:395]: +1 to Alex, I'd say this logic belongs to `RequestTask` (or even to `BackpressureLimiter`?), in that way window-moving logic won't be scattered around the code base and there will be no necessity for `RequestTask` to look inside `RequestTask` for `available_offset`.

by the way,  wouldn't a simplified interface like this suffice for `BackpressureController`?
```rust
impl BackpressureController {
    // Used by `PrefetchGetObject` to report to `BackpressureController` the offset 
    // until which the data was read by the user application.
    pub fn set_next_sequential_read_offset(&mut self, next_sequential_read_offset: usize)
}
```

Ultimately `next_sequential_read_offset` seems to be the only bit of information, which `PrefetchGetObject` should report to `BackpressureLimiter` (via `BackpressureController`) so that the later can make a decision to move the window (apart from the information provided on `RequestTask` instantiation: `preferred_part_size`, `object_size`).

Comment (vladem) [reply mountpoint-s3/src/prefetch.rs:395]: the whole thing becomes much clearer than:
1. we have a `RequestTask` pumping data from CRT to the queue;
2. we have a `PrefetchGetObject` reading that data from the queue when user requests it;
3. `RequestTask` and `PrefetchGetObject` are synchronised through an additional queue, which is used by `PrefetchGetObject` to report when the data is consumed;
4. `RequestTask` stops pumping next data parts from CRT (in `next_read_window_blocking` call) until `PrefetchGetObject` consumes something from the previous part (and reports `next_sequential_read_offset`)

Comment (monthonk) [reply mountpoint-s3/src/prefetch.rs:386]: Just to clarify, the logic here is not for moving the read window but for deciding if we should use bigger window size. Agree that the window moving logic should be in the `RequestTask`.

Comment (vladem) [reply mountpoint-s3/src/prefetch.rs:386]: Looks overcomplicated when we implement window calculation in `PrefetchGetObject`. As I mentioned it needs to fetch an atomic `available_offset` from the `RequestTask` and pass the new window size through the channel to `ClientPartStream`, while it could've passed more basic information in form of consumed offset.

But if you have a strong opinion of why window calculation logic should be in `PrefetchGetObject`, please, share your reasoning!

Comment (monthonk) [reply mountpoint-s3/src/prefetch.rs:387]: Not a really strong opinion, if the logic is in `RequestTask` we still have to pass information about how to scale up the request (like `max_read_window_size` or `sequential_prefetch_multiplier`) to the task as they are currently in the `PrefetcherConfig`. I'm also thinking about if we have some kind of memory limiter in the future, we will also have to push the information/logic about how to scale down to the task.

Comment (vladem) [reply mountpoint-s3/src/prefetch.rs:387]: Ack, some of the information originates from `PrefetcherConfig` and we want to keep it there.

File: mountpoint-s3/src/prefetch.rs [hunk 18]
```diff
@@ -7,14 +7,14 @@
 //! we increase the size of the GetObject requests up to some maximum. If the reader ever makes a
 //! non-sequential read, we abandon the prefetching and start again with the minimum request size.
 
+mod backpressure_controller;
```
Comment (passaro) [inline mountpoint-s3/src/prefetch.rs:10]: Can you update the comment above? It may be a good place to also describe the overall new design and how the various components in this module interact.

**Review by passaro — COMMENTED** [review]
Had another go and left a few comments.

Overall, I'm not too convinced of where the backpressure mechanism and the prefetching logic have ended up, but I don't really have a better proposal. It may only be a matter of better documenting which choices are driven by the current CRT API (and may evolve later), but happy to discuss further.

File: mountpoint-s3/src/prefetch/backpressure_controller.rs [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+/// `wait_for_read_window_increment` to wait for feedback from the consumer. This method
+/// could block when they know that the producer requires read window incrementing.
+///
+/// [BackpressureController] will be given to the consumer side of the object stream.
+/// It can be used anywhere to set preferred read window size for the stream and tell the
+/// producer when its read window should be increased.
+pub fn new_backpressure_controller(
+    preferred_read_window_size: usize,
+    max_read_window_size: usize,
+    read_window_size_multiplier: usize,
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch/backpressure_controller.rs:36]: Yes, it is now. We need to tell it to scale up, but the scaling strategy must be configured when we create `BackpressureController`.

File: mountpoint-s3/src/prefetch/part.rs [hunk 20]
```diff
@@ -21,6 +22,51 @@ impl Part {
...
+    /// Aligning part with the right boundaries will reduce the compute time for validating the checksum at read.
+    pub fn split_to_parts(id: ObjectId, offset: u64, mut body: Bytes, alignment: usize) -> Vec<Self> {
+        let mut curr_offset = offset;
+        let mut parts = Vec::new();
+
+        // First, align the front to part boundaries
+        if offset % alignment as u64 != 0 {
+            let distance_to_align = alignment - (curr_offset % alignment as u64) as usize;
+            let chunk_size = distance_to_align.min(body.len());
+            if chunk_size > 0 {
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch/part.rs:35]: When body is empty, I will make the check explicit.

File: mountpoint-s3/src/prefetch/part_queue.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch/part_queue.rs:115]: I can change it back to `usize`.

File: mountpoint-s3/src/prefetch/task.rs [hunk 21]
```diff
@@ -1,65 +1,98 @@
...
 
 use crate::prefetch::part::Part;
-use crate::prefetch::part_queue::{unbounded_part_queue, PartQueue};
+use crate::prefetch::part_queue::PartQueue;
 use crate::prefetch::PrefetchReadError;
 
+use super::backpressure_controller::BackpressureController;
+use super::part_stream::RequestRange;
+
+pub struct RequestTaskConfig {
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch/task.rs:11]: Will move it to where we define `ObjectPartStream`.

File: mountpoint-s3/src/prefetch/task.rs [hunk 22]
```diff
@@ -1,65 +1,98 @@
...
-            start_offset: offset,
-            total_size: size,
-            part_queue,
+    // Push a given list of parts in front of the part queue
+    pub async fn push_front(&mut self, mut parts: Vec<Part>) -> Result<(), PrefetchReadError<E>> {
+        // Merge all parts into one single part.
+        // This could result in a really big part, but we normally use this only for backward seek
+        // so its size should not be bigger than the prefetcher's `max_backward_seek_distance`.
+        let part = parts.iter_mut().reduce(|acc, e| {
+            acc.extend(e).unwrap();
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch/task.rs:56]: You're right, this should be propagated.

File: mountpoint-s3/src/prefetch/task.rs [hunk 23]
```diff
@@ -1,65 +1,98 @@
...
         self.remaining -= part.len();
+
+        // We read some data out of the part queue so the read window should be moved
+        let next_offset = part.offset() + part.len() as u64;
+        self.backpressure_controller.next_offset_hint(next_offset).await?;
+
+        let remaining_in_queue = self.available_offset().saturating_sub(next_offset) as usize;
+        // If the part queue is empty it means we are reading faster than the task could prefetch,
+        // so we should use larger window for the task.
+        if remaining_in_queue == 0 {
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch/task.rs:79]: Yes, now we have the `BackpressureController` to keep moving the read window with `preferred_read_window_size` which mean data the part queue should always be filled up to this number. if the part queue is not empty it means current `preferred_read_window_size` is large enough to serve the reader. We only want to scale up when the part queue is empty, indicating that the reader is reading faster than we could prefetch.

File: mountpoint-s3/src/prefetch/part_stream.rs [hunk 24]
```diff
@@ -241,7 +216,113 @@ where
...
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    async fn get_from_client(
+        &mut self,
+        initial_read_window_offset: u64,
+    ) -> Result<(), PrefetchReadError<Client::ClientError>> {
+        let bucket = &self.config.bucket;
+        let id = &self.config.object_id;
+
+        // Normally, initial read window size should be very small (~1MB) so that we can serve the first read request as soon as possible,
```
[the remark this answers is not in the document]
Comment (vladem) [reply mountpoint-s3/src/prefetch/part_stream.rs:241]: May be I am missing something here, though. Happy to discuss.

**Review by vladem — COMMENTED** [review]
This POC shows great results, but some aspects of it bother me. Mainly:
1. memory consumption and behaviour change for cache;
2. huge size of the change;
3. amount of code duplication in `part_stream`/`caching_stream`.

I suggest to tackle 3/ first and see if it helps with the size of the change. We may consider enabling backpressure in 2 iterations (only for regular stream at first). Finally we can address issues with caching stream and enable backpressure for it too.

File: mountpoint-s3/src/prefetch/caching_stream.rs [hunk 25]
```diff
@@ -132,56 +153,96 @@ where
...
+            cache_key: cache_key.clone(),
+            preferred_part_size: self.config.preferred_part_size,
+            original_range: range,
+            block_index: block_range.start,
+            block_offset: block_range.start * block_size,
+            buffer: ChecksummedBytes::default(),
+            cache: self.cache.clone(),
+        };
+
+        // Start by issuing the first request that has a range up to initial read window offset.
```
Comment (vladem) [inline mountpoint-s3/src/prefetch/caching_stream.rs:210]: nit: if we had a single function, encapsulating the logic of slitting a request into 2, we won't need to have this duplicated explanation (we could've just use this function in both `ClientPartStream` and `CachingStream`)

**Comment by monthonk** [comment]
I'm going to split this into smaller PRs.
